openpgp/packet/aead_encrypted.go

```diff
@@ -0,0 +1,408 @@
+	writer                io.WriteCloser // 'writer' is a partialLengthWriter
+	writtenEncryptedBytes int
+}
+
+// streamReader reads and decrypts bytes. It caches extra decrypted bytes when
+// necessary, similar to streamWriter.
+type streamReader struct {
+	worker                       // Embedded ciphertext opener
+	reader             io.Reader // 'reader' is a partialLengthReader
+	carry              []byte    // Used to detect last chunk
+	readPlaintextBytes int
+}
+
+func (ae *AEADEncrypted) parse(buf io.Reader) error {
+	ptype, _, contentsReader, err := readHeader(buf)
+	if ptype != packetTypeAEADEncrypted || err != nil {
+		return errors.AEADError("Error reading packet header")
+	}
+	// Information needed for opening chunks
+	header := make([]byte, 5)
+	header[0] = 0x80 | 0x40 | byte(ptype) // Should equal 212
+	if n, err := contentsReader.Read(header[1:5]); err != nil || n < 4 {
+		return errors.AEADError("could not read aead header")
+	}
+	// Read initial nonce
+	nonceLength := (&AEADConfig{mode: AEADMode(header[3])}).NonceLength()
+	initialNonce := make([]byte, nonceLength)
+	if _, err := contentsReader.Read(initialNonce); err != nil {
+		return err
+	}
+	ae.Contents = contentsReader
+	ae.prefix = header
+	ae.initialNonce = initialNonce
+	return nil
+}
+
+// GetStreamReader prepares a worker and returns a ReadCloser from which
+// decrypted bytes can be read (see streamReader.Read()).
+func (ae *AEADEncrypted) GetStreamReader(key []byte) (io.ReadCloser, error) {
+	config := &AEADConfig{
+		version:       ae.prefix[1],
+		cipher:        CipherFunction(ae.prefix[2]),
+		mode:          AEADMode(ae.prefix[3]),
+		chunkSizeByte: byte(ae.prefix[4]),
+	}
+	aead, err := initAlgorithm(key, config)
+	if err != nil {
+		return nil, err
+	}
+	// Carry the first tagLen bytes
+	carry := make([]byte, config.TagLength())
+	n, err := ae.Contents.Read(carry)
+	if err != nil || n < config.TagLength() {
+		return nil, errors.AEADError("Not enough data to decrypt")
+	}
+	return &streamReader{
+		worker: worker{
+			aead:         aead,
+			config:       config,
+			initialNonce: ae.initialNonce,
+			header:       ae.prefix,
+			index:        make([]byte, 8),
+		},
+		reader: ae.Contents,
+		carry:  carry}, nil
+}
+
+// Read decrypts bytes and reads them into dst. It decrypts when necessary and
+// caches extra decrypted bytes. It returns the number of bytes copied into dst
+// and an error.
+func (ar *streamReader) Read(dst []byte) (n int, err error) {
+	if len(dst) == 0 {
+		return 0, errors.AEADError("argument of Read must have positive length")
+	}
+	chunkLen := int(ar.config.ChunkSize())
+	tagLen := ar.config.TagLength()
+	if len(dst) <= len(ar.cache) {
+		n = copy(dst, ar.cache[:len(dst)])
+		ar.cache = ar.cache[n:]
+		return
+	}
+	// Retrieve cached plaintext bytes from previous calls
+	decrypted := ar.cache
+
+	// Decrypt the necessary amount of chunks
+	for i := 0; i <= (len(dst)-len(ar.cache))/chunkLen; i++ {
```

Comment:
Huh, okay, still feels awkward that those functions may or may not deliver complete outputs, for reasons that the caller doesn't know/can't access (in this case, the chunk limits...). 
Actually, `ReadFull` iterates on `Read` calls until it gets what it needs (see `io.go ReadAtLeast`).

Reply:
The idea is: imagine we want to display a message progressively (i.e. streaming) on a slow network connection. Let's say the ethernet packet size is 1kb, the chunk size is 2kb, and the application would ideally like to display the message in as large chunks as possible for performance, but can do small ones if necessary. Let's say it passes a 4kb slice to `Read`. As written, your code would wait on the network for 4 packets to come in, decode two chunks, and return them. This means that the first chunk is shown to the user later than necessary. So instead, we only want to wait on the network for 2 packets, decode one chunk, and return it.

If the `Read` function didn't exist and there was only `ReadFull`, the application would have to know about ethernet packet sizes, and chunk sizes, and then pass a 1kb slice in order to receive data as soon as possible (also hampering performance on faster networks). So that's why the `Read` interface is defined as it is, in order to allow the caller to not care about things that it doesn't know/can't access :)